src/ingest/SpecRepo.cpp

```diff
@@ -147,22 +173,21 @@ void genSpecs4Roll(const std::string& version,
     // making a s3 fs with given host
     auto fs = nebula::storage::makeFS("s3", sourceInfo.host);
 
+    // exact macro pattern type
+    auto pt = nebula::meta::extractPatternMacro(table->timeSpec.pattern);
+
     // list all objects/files from given path
     // A roll spec will cover X days given table location of source data
     const auto now = Evidence::now();
     const auto max_days = table->max_hr / 24;
-    for (size_t i = 0; i <= max_days; ++i) {
-      // we only provide single macro for now
-      auto timeValue = now - i * DAY_SECONDS;
-      auto path = fmt::format(
-        sourceInfo.path, fmt::arg("date", Evidence::fmt_ymd_dash(timeValue)));
-      auto files = fs->list(path);
-      genSpecPerFile(table, version, files, specs, timeValue);
-    }
 
+    // earliest time in second to process in ascending order
+    long cutOffTime = now - table->max_hr * nebula::meta::unitInSeconds.at(nebula::meta::PatternMacro::HOUR);
+
+    // TODO(chenqin): don't support other macro other than dt=date/hr=hour/mi=minute/se=second yet. chenqin
+    genSpec(max_days + 2, nebula::meta::PatternMacro::DATE, pt, now, sourceInfo.path, cutOffTime, version, table, specs);
```

Comment:
why max_days  +2? 
btw, could you rename `max_days` to `maxDays`, it was my mistake?

Reply:
good catch, refactored a bit to use end index explicitly.